secure/net/src/network_controller/outbound_handler.rs

```diff
@@ -29,84 +30,126 @@ impl OutboundHandler {
             inbound_handler,
         }
     }
 
     pub fn register_handler(
-        &self,
+        &mut self,
         message_type: String,
         remote_addr: SocketAddr,
         receiver: Receiver<Message>,
     ) {
-        // Create a remote client if it doesn't exist
-        self.network_clients
-            .lock()
-            .unwrap()
-            .entry(remote_addr)
-            .or_insert_with(|| NetworkClient::new(message_type.clone(), remote_addr, 5000));
-        let mut handlers = self.handlers.lock().unwrap();
-        handlers.push((receiver, remote_addr, MessageType::new(message_type)));
+        self.remote_addresses.insert(remote_addr);
+        self.handlers
+            .push((receiver, remote_addr, MessageType::new(message_type)));
     }
 
-    pub fn start(&mut self) {
-        let outbound_handlers = self.handlers.clone();
+    pub fn start(&mut self, rt: &Runtime) -> Option<Sender<Message>> {
+        if self.handlers.is_empty() {
+            return None;
+        }
+
+        // Register a signal handler to stop the outbound task
+        let (stop_signal_tx, stop_signal_rx) = unbounded();
+        self.handlers.push((
+            stop_signal_rx,
+            self.address,
+            MessageType::new("stop_task".to_string()),
+        ));
+
+        // Create a grpc client for each remote address
+        let mut grpc_clients: HashMap<SocketAddr, RemoteExecutionClientWrapper> = HashMap::new();
+        self.remote_addresses.iter().for_each(|remote_addr| {
+            grpc_clients.insert(
+                *remote_addr,
+                RemoteExecutionClientWrapper::new(rt, *remote_addr),
+            );
+        });
+
+        // Prepare for objects to be moved into the async block (&mut self cannot be moved into the
+        // async block)
         let address = self.address;
-        let network_clients = self.network_clients.clone();
-        let thread_name = format!("{}_network_outbound_handler", self.service);
-        let builder = thread::Builder::new().name(thread_name);
         let inbound_handler = self.inbound_handler.clone();
-        builder
-            .spawn(move || loop {
-                Self::process_one_outgoing_message(
-                    outbound_handlers.clone(),
-                    network_clients.clone(),
-                    &address,
-                    inbound_handler.clone(),
-                )
-            })
-            .expect("Failed to spawn outbound handler thread");
+        // Moving the handlers out of self is fine because once 'start()' is called we do not intend
+        // to register any more handlers. A reference count like Arc<Mutex> has issues of being
+        // used across sync and async boundaries, and also not the most efficient because we pay
+        // the cost of the mutex when there is no contention
+        let outbound_handlers = mem::take(self.handlers.as_mut());
+
+        // TODO: Consider using multiple tasks for outbound handlers
```

Comment:
What would it take to multi-thread this? I am afraid, without processing messages in multiple threads, this will be a chokepoint in remote execution environment.

Reply:
I suppose minimal overhead. Just the logic of splitting the outbound_handlers across the threads/tasks